### DISCUSSION TITLE
Emacs setup for vue + typescript (volar + eglot + web-mode)

### DISCUSSION THREAD

**Discussion by aadi58002** [joaotavora/eglot#1184] [Q&A] [answered]
Hello there,I searched through the eglot issues and discussion and most i could find was this discussion for a eglot recipe for volarhttps://github.com/joaotavora/eglot/discussions/1058

but the server connects and starts but i don't get any completion.

Reddit Question -> https://www.reddit.com/r/emacs/comments/11svcvj/emacs_setup_for_vue_typescript_volar_eglot_webmode/
Eglot logs https://pastebin.com/nttpMqPa

**Comment by rsierkov** [comment, marked as answer]
Had the same issue and this is what worked for me:

```
;; web-mode setup
(define-derived-mode vue-mode web-mode "Vue")
(add-to-list 'auto-mode-alist '("\\.vue\\'" . vue-mode))

(defun vue-eglot-init-options ()
             (let ((tsdk-path (expand-file-name
                               "lib"
                               (shell-command-to-string "npm list --global --parseable typescript | head -n1 | tr -d \"\n\""))))
               `(:typescript (:tsdk ,tsdk-path
                              :languageFeatures (:completion
                                                 (:defaultTagNameCase "both"
                                                  :defaultAttrNameCase "kebabCase"
                                                  :getDocumentNameCasesRequest nil
                                                  :getDocumentSelectionRequest nil)
                                                 :diagnostics
                                                 (:getDocumentVersionRequest nil))
                              :documentFeatures (:documentFormatting
                                                 (:defaultPrintWidth 100
                                                  :getDocumentPrintWidthRequest nil)
                                                 :documentSymbol t
                                                 :documentColor t)))))

;; Volar
(add-to-list 'eglot-server-programs
                        `(vue-mode . ("vue-language-server" "--stdio" :initializationOptions ,(vue-eglot-init-options))))
```

**Reply by pbgc** [reply]
This was not working for me.. had to change:

`(shell-command-to-string "npm list --global --parseable typescript | head -n1 | tr -d \"\n\"")`

to

`(string-trim-right (shell-command-to-string "npm list --global --parseable typescript | head -n1"))`

**Comment by trev-dev** [comment]
Has anyone gotten this to actually work?  It can't seem to find the "tdsk" root in my case.  I get an error:
`error in process filter: [eglot] -32603: Request initialize failed with message: Can't find typescript.js or tsserverlibrary.js in undefined`

```elisp
(defun eglot-volar-find-typescript ()
  "Get the global npm typescript location."
  (string-trim-right
   (shell-command-to-string (concat "npm list --global --parseable typescript"
                                    " | head -n1"))))

(defun eglot-volar-init-options ()
  "Get the vue-language-server initOptions."
  (let ((tdsk-path (expand-file-name "lib" (eglot-volar-find-typescript))))
    `(:typescript (:tdsk ,tdsk-path
                   :languageFeatures (:completion
                                      (:defaultTagNameCase "both"
                                       :defaultAttrNameCase "kebabCase"
                                       :getDocumentNameCasesRequest nil
                                       :getDocumentSelectionRequest nil)
                                   :diagnostics
                                   (:getDocumentVersionRequest nil))
                   :documentFeatures (:documentFormatting
                                      (:defaultPrintWidth 100
                                       :getDocumentPrintWidthRequest nil)
                                      :documentSymbol t
                                      :documentColor t)))))

(with-eval-after-load 'eglot
  (add-to-list 'eglot-server-programs
               `(vue-mode . ("vue-language-server"
                             "--stdio"
                             :initializationOptions
                             ,(eglot-volar-init-options))))
```

**Comment by pbgc** [comment]
@trev-dev i give you my vue setup that works great (for me). Maybe it can help:


```
(define-derived-mode pbgc-vue-mode web-mode "pbVue"
  "A major mode derived from web-mode, for editing .vue files with LSP support.")


(add-to-list 'auto-mode-alist '("\\.vue\\'" . pbgc-vue-mode))

(defun pbgc-vue-custom ()
  (flycheck-mode t)
  (eglot-ensure))
(add-hook 'pbgc-vue-mode-hook 'pbgc-vue-custom)

(defun vue-eglot-init-options ()
             (let ((tsdk-path (expand-file-name
                               "lib"
                               (string-trim-right (shell-command-to-string "npm list --global --parseable typescript | head -n1"))
                               )))
               `(:typescript (:tsdk ,tsdk-path
                              :languageFeatures (:completion
                                                 (:defaultTagNameCase "both"
                                                  :defaultAttrNameCase "kebabCase"
                                                  :getDocumentNameCasesRequest nil
                                                  :getDocumentSelectionRequest nil)
                                                 :diagnostics
                                                 (:getDocumentVersionRequest nil))
                              :documentFeatures (:documentFormatting
                                                 (:defaultPrintWidth 100
                                                  :getDocumentPrintWidthRequest nil)
                                                 :documentSymbol t
                                                 :documentColor t)))))

;; Volar
(add-to-list 'eglot-server-programs
                        `(pbgc-vue-mode . ("vue-language-server" "--stdio" :initializationOptions ,(vue-eglot-init-options))))
```

**Reply by trev-dev** [reply]
It helps in the sense that it looks a lot like mine.  If I inspect `eglot-server-programs`, I see this as the first member of my list:

```elisp
(vue-mode "vue-language-server" "--stdio"
          :initializationOptions (:typescript
                                  (:tdsk "/home/trev/.local/node/lib/node_modules/typescript/lib"
                                         :languageFeatures
                                         (:completion
                                          (:defaultTagNameCase "both"
                                           :defaultAttrNameCase "kebabCase"
                                           :getDocumentNameCasesRequest nil
                                           :getDocumentSelectionRequest nil)
                                          :diagnostics (:getDocumentVersionRequest nil))
                                         :documentFeatures
                                         (:documentFormatting
                                          (:defaultPrintWidth 100 :getDocumentPrintWidthRequest nil)
                                          :documentSymbol t :documentColor t))))
```

The "missing" files `typescript.js or tsserverlibrary.js` are indeed at `"/home/trev/.local/node/lib/node_modules/typescript/lib"`

**Reply by trev-dev** [reply]
I got it.  Eglot was timing out too quickly.  I set `eglot-connect-timeout` to 120 and it works.

**Comment by unhammer** [comment]
For those installing vue-language-server from nix:

```emacs-lisp
(define-derived-mode vue-mode web-mode "Vue")
(add-to-list 'auto-mode-alist '("\\.vue\\'" . vue-mode))
(defun vue-eglot-init-options ()
  "Set SDK path and default options."
  (let ((tsdk-path (expand-file-name
                    "lib/node_modules/typescript/lib/"
                    (shell-command-to-string
                     (string-join '("nix-store  --query --references `which vue-language-server`"
                                    "xargs -n1 nix-store -q --referrers"
                                    "grep typescript-language-server"
                                    "tr -d '\n'")
                                  " | ")))))
    `(:typescript (:tsdk ,tsdk-path
                         :languageFeatures (:completion
                                            (:defaultTagNameCase "both"
                                                                 :defaultAttrNameCase "kebabCase"
                                                                 :getDocumentNameCasesRequest nil
                                                                 :getDocumentSelectionRequest nil)
                                            :diagnostics
                                            (:getDocumentVersionRequest nil))
                         :documentFeatures (:documentFormatting
                                            (:defaultPrintWidth 100
                                                                :getDocumentPrintWidthRequest nil)
                                            :documentSymbol t
                                            :documentColor t)))))
(add-to-list 'eglot-server-programs ;; nix-env -iA nixpkgs.nodePackages.volar
             `(vue-mode . ("vue-language-server" "--stdio" :initializationOptions ,(vue-eglot-init-options))))
```

**Reply by joaotavora** [reply]
Nice work Looks like you guys should get together and write a `vue-mode.el` :-)

**Reply by garyo** [reply]
I'm still struggling with this too. A few diffs for me, but still not working:
```
    (let ((tsdk-path (expand-file-name
                      "../global/node_modules/typescript/lib"
                      (string-trim-right (shell-command-to-string "yarn global bin"))
                      )))
```
That sets `tsdk-path` to a lib folder that contains a `typescript.js` but I don't know if it's right. `tsc --version` shows 5.5.4 for me, and `package.json` in that folder agrees, 5.5.4. But when starting the language server, I see this in the stderr log:
```
[stderr]  No typescript found, vue-language-server will not work.
...
[jsonrpc] e[10:30:23.241] --> client/registerCapability[1] {"jsonrpc":"2.0","id":1,"error":{"code":-32603,"message":"Internal error"}}
```
(note that I start the language server as `"bun" "x" "@volar/vue-language-server" "--stdio"` and that is working.)

I think what's happening is my `initializationOptions` are not getting passed to the server. Here's the initial message passed to the server:
```
[jsonrpc] D[10:30:22.904] Running language server: bun x @volar/vue-language-server --stdio
[jsonrpc] e[10:30:22.906] --> initialize[1] {"jsonrpc":"2.0","id":1,"method":"initialize","params":{"processId":14137,"clientInfo":{"name":"Eglot","version":"1.17"},"rootPath":"/Users/garyo/astrid/tutor-site/hugorcd-canvas/","rootUri":"file:///Users/garyo/astrid/tutor-site/hugorcd-canvas",
"initializationOptions":{},
"capabilities":{"workspace":{"applyEdit":true,"executeCommand":{"dynamicRegistration":false},"workspaceEdit":{"documentChanges":true},"didChangeWatchedFiles":{"dynamicRegistration":true},"symbol":{"dynamicRegistration":false},"configuration":true,"workspaceFolders":true},"textDocument":{"synchronization":{"dynamicRegistration":false,"willSave":true,"willSaveWaitUntil":true,"didSave":true},"completion":{"dynamicRegistration":false,"completionItem":{"snippetSupport":true,"deprecatedSupport":true,"resolveSupport":{"properties":["documentation","details","additionalTextEdits"]},"tagSupport":{"valueSet":[1]}},"contextSupport":true},"hover":{"dynamicRegistration":false,"contentFormat":["markdown","plaintext"]},"signatureHelp":{"dynamicRegistration":false,"signatureInformation":{"parameterInformation":{"labelOffsetSupport":true},"documentationFormat":["markdown","plaintext"],"activeParameterSupport":true}},"references":{"dynamicRegistration":false},"definition":{"dynamicRegistration":false,"linkSupport":true},"declaration":{"dynamicRegistration":false,"linkSupport":true},"implementation":{"dynamicRegistration":false,"linkSupport":true},"typeDefinition":{"dynamicRegistration":false,"linkSupport":true},"documentSymbol":{"dynamicRegistration":false,"hierarchicalDocumentSymbolSupport":true,"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]}},"documentHighlight":{"dynamicRegistration":false},"codeAction":{"dynamicRegistration":false,"resolveSupport":{"properties":["edit","command"]},"dataSupport":true,"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite","source","source.organizeImports"]}},"isPreferredSupport":true},"formatting":{"dynamicRegistration":false},"rangeFormatting":{"dynamicRegistration":false},"rename":{"dynamicRegistration":false},"inlayHint":{"dynamicRegistration":false},"publishDiagnostics":{"relatedInformation":false,"codeDescriptionSupport":false,"tagSupport":{"valueSet":[1,2]}}},"window":{"showDocument":{"support":true},"workDoneProgress":true},"general":{"positionEncodings":["utf-32","utf-8","utf-16"]},"experimental":{}},"workspaceFolders":[{"uri":"file:///Users/garyo/astrid/tutor-site/hugorcd-canvas","name":"~/astrid/tutor-site/hugorcd-canvas/"}]}}
```
but my `eglot-server-programs` var looks correct: 
```
ELISP> eglot-server-programs
...
 (vue-mode "bun" "x" "vue-language-server" "--stdio"
           :initializationOptions
           (:typescript
            (:tsdk
             "/Users/garyo/.config/yarn/global/node_modules/typescript/lib"
             :languageFeatures
             (:completion
              (:defaultTagNameCase "both" :defaultAttrNameCase
                                   "kebabCase"
                                   :getDocumentNameCasesRequest nil
                                   :getDocumentSelectionRequest nil)
              :diagnostics (:getDocumentVersionRequest nil))
             :documentFeatures
             (:documentFormatting
              (:defaultPrintWidth 100 :getDocumentPrintWidthRequest
                                  nil)
              :documentSymbol t :documentColor t))))
...
```

**Reply by garyo** [reply]
What would help me debug this is understanding how the init options get sent to the server. I see some cl-defgenerics in the source, but have no idea how to edebug them or add a `(message ...)`.

**Reply by pbgc** [reply]
@garyo currently I have Volar (last version) working without any problem with this conf:


```
(define-derived-mode pbgc-vue-mode web-mode "pbVue"
  "A major mode derived from web-mode, for editing .vue files with LSP support.")

(add-to-list 'auto-mode-alist '("\\.vue\\'" . pbgc-vue-mode))

(defun pbgc-vue-custom ()
  (flycheck-mode t)
  (corfu-mode)
  (eglot-ensure))
(add-hook 'pbgc-vue-mode-hook 'pbgc-vue-custom)

(defun vue-eglot-init-options ()
             (let ((tsdk-path (expand-file-name
                               "lib"
                               (string-trim-right (shell-command-to-string "npm list --global --parseable typescript | head -n1")))))
               `(:typescript (:tsdk ,tsdk-path)
                 :vue (:hybridMode :json-false))))


;; Volar

(put 'pbgc-vue-mode 'eglot-language-id "vue")

(add-to-list 'eglot-server-programs
             `(pbgc-vue-mode . ("vue-language-server" "--stdio" :initializationOptions ,(vue-eglot-init-options))))
```

**Reply by garyo** [reply]
I'm doing pretty much that, but the init options aren't getting passed to the server. Can you check your eglot log to see if yours are getting passed?

**Reply by pbgc** [reply]
@garyo I'm 100% sure that the init options are being passed. I already checked that in the past .. but I don't need to .. because if the typescript option and and hybridMode: false were not being passed  ... it would not work at all! (without setting hybridMode to false I would need to have another separated Language Server for typescript ... and eglot cannot handle more than a Language Server at the same time)

**Reply by pbgc** [reply]
@garyo looking at your settings .. you are missing the:


**:vue (:hybridMode :json-false)**

and 

**(put 'pbgc-vue-mode 'eglot-language-id "vue")**


Comment you code and try to use mine above .. to check if it will work (I think it will) .. or there's another problem.
I'm using: @vue/language-server@2.0.29

**Reply by garyo** [reply]
Indeed -- I had a couple of errors in my code (including my file wasn't getting set into the proper major mode), so it wasn't getting the options passed properly. Got it going now. 
I do still get "[eglot] (warning) Server tried to register unsupported capability `workspace/didChangeWorkspaceFolders'" and it's not giving me any completions or results yet, but at least the server seems to stay running.

**Reply by pbgc** [reply]
@garyo just ignore the warning (I get it too..and can be ignored. You could try to disable that registration attempt by eglot .. but just ignore it!).
If you don't get any completions ... you don't have corfu-mode or company-mode enabled on the buffer!

**Comment by ghost** [comment]
fyi treesit-auto recently merged support for downloading vue grammars for tree-sitter: https://github.com/renzmann/treesit-auto/pull/86 so now theoretically you could download vue-ts-mode to get tree-sitter support instead of web-mode

Run `(treesit-install-all)` after installing the treesit-auto pkg to install the grammars so that vue-ts-mode can work.

I'm using Elpaca with use-package here
```elisp
(use-package treesit-auto
    :ensure t
    :config
    (setq treesit-auto-install t)
    (treesit-auto-add-to-auto-mode-alist 'all)
    (global-treesit-auto-mode))

(use-package vue-ts-mode
    :ensure (:host github :repo "8uff3r/vue-ts-mode"))

  (use-package eglot
    :preface
    (defun vue-eglot-init-options ()
      (let ((tsdk-path (expand-file-name "lib" (string-trim-right (shell-command-to-string "NODE_NO_WARNINGS=1 npm list --global --parseable typescript | head -n1")))))
	`(:typescript (:tsdk ,tsdk-path) :vue (:hybridMode :json-false))))
    :config
    (add-to-list 'eglot-server-programs
    		 `(vue-ts-mode . ("vue-language-server" "--stdio" :initializationOptions ,(vue-eglot-init-options))))
    (setq eglot-ignored-server-capabilities '(:documentFormatting :documentFormattingProvider)))

```

Make sure you install the @vue/language-server package and not vue-language-server

EDIT: To below commenter, this was probably because of lack of `:hybridMode :json-false`

**Reply by pedroprieto** [reply]
I have tried it with no luck. I have modified the config so that eglot sends `languageid = vue` instead of `vue-ts`. However, I only get emmet completions (html and css), but no JavaScript integration. Does eglot support it?

**Comment by joaotavora** [comment]
The registration attempt is not by Eglot. The message is pretty clear about
that

**Reply by garyo** [reply]
Yeah, agreed. I'm just trying to figure out why the hover and highlight events just return nothing. Something server-side I'm sure.

```
[jsonrpc] e[10:58:47.782] --> textDocument/didOpen {"jsonrpc":"2.0","method":"textDocument/didOpen","params":{"textDocument":{"uri":"file:///Users/garyo/astrid/tutor-site/hugorcd-canvas/app/pages/articles/%5B...slug%5D.vue","version":0,"languageId":"gco-vue","text":"<script lang=\"ts\" setup>\nimport { Toaster } from 'vue-sonner'\n\nconst { t, locale } = useI18n()\n\nconst route = useRoute()\nconst { data: page } = await useAsyncData(`${route.path}`, () => queryContent(route.path).findOne())\n\nif (!page.value) {\n  throw createError({ statusCode: 404, statusMessage: 'Page not found' })\n}\n\nuseContentHead(page.value)\n\nconst runtimeConfig = useRunti...
...
[jsonrpc] e[10:58:51.389] --> textDocument/hover[3] {"jsonrpc":"2.0","id":3,"method":"textDocument/hover","params":{"textDocument":{"uri":"file:///Users/garyo/astrid/tutor-site/hugorcd-canvas/app/pages/articles/%5B...slug%5D.vue"},"position":{"line":21,"character":9}}}
[jsonrpc] e[10:58:51.389] --> textDocument/documentHighlight[4] {"jsonrpc":"2.0","id":4,"method":"textDocument/documentHighlight","params":{"textDocument":{"uri":"file:///Users/garyo/astrid/tutor-site/hugorcd-canvas/app/pages/articles/%5B...slug%5D.vue"},"position":{"line":21,"character":9}}}
[jsonrpc] e[10:58:51.397] <-- textDocument/hover[3] {"jsonrpc":"2.0","id":3,"result":null}
[jsonrpc] e[10:58:51.398] <-- textDocument/documentHighlight[4] {"jsonrpc":"2.0","id":4,"result":[]}
[jsonrpc] e[10:58:54.494] --> textDocument/hover[5] {"jsonrpc":"2.0","id":5,"method":"textDocument/hover","params":{"textDocument":{"uri":"file:///Users/garyo/astrid/tutor-site/hugorcd-canvas/app/pages/articles/%5B...slug%5D.vue"},"position":{"line":22,"character":8}}}
[jsonrpc] e[10:58:54.494] --> textDocument/documentHighlight[6] {"jsonrpc":"2.0","id":6,"method":"textDocument/documentHighlight","params":{"textDocument":{"uri":"file:///Users/garyo/astrid/tutor-site/hugorcd-canvas/app/pages/articles/%5B...slug%5D.vue"},"position":{"line":22,"character":8}}}
[jsonrpc] e[10:58:54.505] <-- textDocument/hover[5] {"jsonrpc":"2.0","id":5,"result":null}
[jsonrpc] e[10:58:54.505] <-- textDocument/documentHighlight[6] {"jsonrpc":"2.0","id":6,"result":[]}
[jsonrpc] e[10:58:55.503] --> textDocument/hover[7] {"jsonrpc":"2.0","id":7,"method":"textDocument/hover","params":{"textDocument":{"uri":"file:///Users/garyo/astrid/tutor-site/hugorcd-canvas/app/pages/articles/%5B...slug%5D.vue"},"position":{"line":36,"character":9}}}
[jsonrpc] e[10:58:55.503] --> textDocument/documentHighlight[8] {"jsonrpc":"2.0","id":8,"method":"textDocument/documentHighlight","params":{"textDocument":{"uri":"file:///Users/garyo/astrid/tutor-site/hugorcd-canvas/app/pages/articles/%5B...slug%5D.vue"},"position":{"line":36,"character":9}}}
[jsonrpc] e[10:58:55.514] <-- textDocument/hover[7] {"jsonrpc":"2.0","id":7,"result":null}
[jsonrpc] e[10:58:55.515] <-- textDocument/documentHighlight[8] {"jsonrpc":"2.0","id":8,"result":[]}
```

**Reply by pbgc** [reply]
@joaotavora  right! Sorry

**Comment by ghost** [comment]
In case it's helpful, the vue-eglot-init-options config [above](https://github.com/joaotavora/eglot/discussions/1184#discussioncomment-6492935) has a bug, which prevents the server from giving diagnostics in the `<script>` section of a Vue SFC. Here's a correct version:
```
  (defun vue-eglot-init-options ()
    (let ((tsdk-path (typescript-lib-dir)))
      `(:typescript (:tsdk ,tsdk-path)
                    :vue (:hybridMode :json-false)
                    :languageFeatures (:completion
                                       (:defaultTagNameCase "both"
                                                            :defaultAttrNameCase "kebabCase"
                                                            :getDocumentNameCasesRequest nil
                                                            :getDocumentSelectionRequest nil)
                                       :diagnostics
                                       (:getDocumentVersionRequest nil))
                    :documentFeatures (:documentFormatting
                                       (:defaultPrintWidth 100
                                                           :getDocumentPrintWidthRequest nil)
                                       :documentSymbol t
                                       :documentColor t))))

  )
```
The change is that `:vue` and `:languageFeatures` etc. are at the *same* JSON level as `:typescript`, not a sub-object of it (so just moving a close-paren, really). The resulting JSON sent to the server looks like this:
```json
    "initializationOptions": {
      "typescript": {
        "tsdk": "/Users/garyo/.config/yarn/global/node_modules/typescript/lib"
      },
      "vue": {
        "hybridMode": false
      },
      "languageFeatures": {
        "completion": {
          "defaultTagNameCase": "both",
          "defaultAttrNameCase": "kebabCase",
          "getDocumentNameCasesRequest": null,
          "getDocumentSelectionRequest": null
        },
        "diagnostics": {
          "getDocumentVersionRequest": null
        }
      },
      "documentFeatures": {
        "documentFormatting": {
          "defaultPrintWidth": 100,
          "getDocumentPrintWidthRequest": null
        },
        "documentSymbol": true,
        "documentColor": true
      }
    },
 ```

**Reply by pbgc** [reply]
@garyo the config you are pointing does not have any :vue section ... it's an older version before Volar introducing Hybrid Mode ..
with the one I sent after (https://github.com/joaotavora/eglot/discussions/1184#discussioncomment-10444494) I get diagnostics..

**Reply by garyo** [reply]
Yes indeed, you're right, that one you linked to has the fix, close-paren after `,tsdk-path`. So:`(:typescript (:tsdk ,tsdk-path) :vue (:hybridMode :json-false)) ...`
I'm using a slightly-modified globally-installed @vue/language-server@2.0.29 (had to clean up a couple of its dependencies in its package.json due to [recent volar breaking changes](https://github.com/vuejs/language-tools/issues/4706)) and it works with my config but not with yours (with yours I get no docs in the typescript part). There certainly are lots of moving parts though!

**Reply by aghull** [reply]
I haven't been able to get this to work with eglot after much effort. No matter what I try the language server gives blank replies to any `textDocument/documentHighlight` or `textDocument/hover` requests. I don't know how to continue troubleshooting. LSP mode, otoh, works out of the box with the exact same language server install and seemingly the exact same requests. What could I be doing wrong?

Also why would you be looking for a global typescript lib for the tsdk instead of the project root as [LSP](https://github.com/emacs-lsp/lsp-mode/blob/master/clients/lsp-volar.el#L86) does?

**Reply by aghull** [reply]
I was also missing `:hybridMode` not realizing this defaults to true and getting lost in the various snippets posted here. I'm still don't think passing the global typescript as tsdk is the best approach. I wish this would just get PR'd and merged to avoid the hassle.

**Comment by ghost** [comment]
I'd like to add that on Node 23 the `shell-command-to-string` call should probably be

```diff
- npm list --global --parseable typescript | head -n1
+ NODE_NO_WARNINGS=1 npm list --global --parseable typescript | head -n1
```

because the first version (without NODE_NO_WARNINGS) returns (on Node 23):

```
(node:12651) ExperimentalWarning: CommonJS module /opt/homebrew/lib/node_modules/npm/node_modules/debug/src/node.js is loading ES Module /opt/homebrew/lib/node_modules/npm/node_modules/supports-color/index.js using require().
Support for loading ES Module in require() is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
/opt/homebrew/lib/node_modules/typescript
```

so due to the `head -n1` the `(node:12651) [...]` line gets taken as the typescript path, which makes the LSP fail because it's not a valid path.

**Comment by ghost** [comment]
EDIT: doesn't work, click "show 6 previous replies" and see first reply below

~With @vue/language-server v3 tsdk.path is no longer required https://github.com/vuejs/language-tools/discussions/5456~

<details>
<summary>code block that doesn't work</summary>
```elisp
;; make sure you're using one of the vue-ts-mode or web-mode and change the `vue-ts-mode` to something else
(add-to-list 'eglot-server-programs
	'(vue-ts-mode . ("vue-language-server" "--stdio" :initializationOptions '(:vue (:hybridMode :json-false)))))
```
</details>

**Reply by pbgc** [reply]
in V3 hybrid mode was removed ... and a different lsp for typescript must be used .. so I don't think this will work.
As eglot does not support running more than 1 lsp at the same time ... only by using something like lspx this can be made to work, eventually.
I'm still using V2 because of that ...

**Reply by trev-dev** [reply]
> in V3 hybrid mode was removed ... and a different lsp for typescript must be used .. so I don't think this will work. As eglot does not support running more than 1 lsp at the same time ... only by using something like lspx this can be made to work, eventually. I'm still using V2 because of that ...

I'm just sitting in the "why even bother" camp with `.vue` files. If they're just going to break things upstream every 6-8 months I'd rather just use grep. Edit: at least `flymake-eslint` works

**Reply by pbgc** [reply]
> I'm just sitting in the "why even bother" camp with `.vue` files. If they're just going to break things upstream every 6-8 months I'd rather just use grep. Edit: at least `flymake-eslint` works

I really don't think vue-language-server is breaking things every 6-8 months! This change was LONG announced and hybrid mode was already deprecated!! And the change is for a lot better! (in terms of performance and simplicity)
The problem here is only the fact that eglot does not support, by option, more than 1 lsp at the same time .. or it would be very straightforward to change! 
It's working great on NeoVim, Zed, VSCode and Emacs lsp mode will support it (if it's not doing it now already)

**Reply by trev-dev** [reply]
> It's working great on NeoVim, Zed, VSCode and Emacs lsp mode will support it (if it's not doing it now already)

As of 2 days ago it was broken in `lsp-mode`.

I can see it working in Neovim but politely, I've had to update some LSP client in one editor or another for Vue more times than I care for. Zed and VSCode have an emphasis on convenience, so good for them for keeping up. I'm just grateful that I know Vue well enough by now to not necessarily depend on it.

**Reply by pbgc** [reply]
> I've had to update some LSP client in one editor or another for Vue more times than I care for

True indeed! but I think it's heading in the right direction now (or at least  I hope).

**Reply by idlip** [reply]
@pbgc In other thread you said you got vue lsp working with `lspx`
Can you just describe how to use?
I tried using it, but I get various timed out error

`jsonrpc-request: jsonrpc-error: "request id=7 failed:", (jsonrpc-error-message . "Timed out")`
Everytime id=n will be diff number

Thanks

**Reply by pbgc** [reply]
@idlip I used lspx to be able to have the tailwind lsp with vue (this conf only works with vue language server 2):

```
(define-derived-mode pbgc-vue-mode web-mode "pbVue"
  "A major mode derived from web-mode, for editing .vue files with LSP support.")

(add-to-list 'auto-mode-alist '("\\.vue\\'" . pbgc-vue-mode))

(defun pbgc-vue-custom ()
  (flycheck-mode t)
  (corfu-mode)
  (eglot-ensure))
(add-hook 'pbgc-vue-mode-hook 'pbgc-vue-custom)

(defun vue-eglot-init-options2 ()
             (let ((tsdk-path (expand-file-name
                               "lib"
                               (string-trim-right (shell-command-to-string "npm list --global --parseable typescript | head -n1")))))
               `(:typescript (:tsdk ,tsdk-path)
                             :vue (:hybridMode :json-false))))

(put 'pbgc-vue-mode 'eglot-language-id "vue")

(add-to-list 'eglot-server-programs `(pbgc-vue-mode . ("lspx" "--lsp" "tailwindcss-language-server --stdio" "--lsp" "vue-language-server --stdio" :initializationOptions ,(vue-eglot-init-options2))))
```

**Reply by idlip** [reply]
ah, I was trying with vue-language-server version >3

**Reply by ParsaJR** [reply]
> With @vue/language-server v3 tsdk.path is no longer required [vuejs/language-tools#5456](https://github.com/vuejs/language-tools/discussions/5456)
> 
> ```emacs-lisp
> ;; make sure you're using one of the vue-ts-mode or web-mode and change the `vue-ts-mode` to something else
> (add-to-list 'eglot-server-programs
> 	'(vue-ts-mode . ("vue-language-server" "--stdio" :initializationOptions '(:vue (:hybridMode :json-false)))))
> ```

Is it actually working for you?
I'm trying to setup the `vue-language-server` with eglot. (And if i make it work, proceeding with [rassumfrassum](https://github.com/joaotavora/rassumfrassum) is the main goal.)

it doesn't give me any clue whats wrong with the vue's language server. After some time, eglot client reports a bunch of timeout requests:
```
[jsonrpc] i[12:59:30.384] [2] timed-out request ':textDocument/inlayHint'
[jsonrpc] i[12:59:30.885] [3] timed-out request ':textDocument/codeAction'
[jsonrpc] i[12:59:30.885] [4] timed-out request ':textDocument/documentHighlight'
[jsonrpc] i[12:59:30.885] [5] timed-out request ':textDocument/hover'
[jsonrpc] i[12:59:37.594] [6] timed-out request ':textDocument/codeAction'
[jsonrpc] i[12:59:37.594] [7] timed-out request ':textDocument/documentHighlight'
[jsonrpc] i[12:59:37.594] [8] timed-out request ':textDocument/hover'
```

Now, the thing is, i think something is missing in my current config:
```
(with-eval-after-load 'eglot
  (add-to-list 'eglot-server-programs
               '(vue-ts-mode . ("vue-language-server" "--stdio"))))

```
Compare this with the working Neovim's lsp-client configuration for 'vue-language-server`:
```
			-- VUE_LS
			local vue_language_server_path = vim.fn.stdpath('data')
				.. "/mason/packages"
				.. "/vue-language-server"
				.. "/node_modules/@vue/language-server"

			local tsserver_filetypes = { "typescript", "javascript", "javascriptreact", "typescriptreact", "vue" }
			local vue_plugin = {
				name = "@vue/typescript-plugin",
				location = vue_language_server_path,
				languages = { "vue" },
				configNamespace = "typescript",
			}

			local ts_ls_config = {
				init_options = {
					plugins = {
						vue_plugin,
					},
				},
				filetypes = tsserver_filetypes,
			}

			local vue_ls_config = {}

			vim.lsp.config("vue_ls", vue_ls_config)
			vim.lsp.config("ts_ls", ts_ls_config)
			vim.lsp.enable({ "ts_ls", "vue_ls" })
```

`vue-language-server` needs to be combined with `ts_ls` in some way? Or it's not required?

The common teammate  of vue.js, `tailwindcss-lanaguage-server` is working just fine. I just dont know what is wrong with the vue's toolchain.

**Reply by pbgc** [reply]
@ParsaJR Only V2 works. V3 don't work .. it needs extra suport like you can read here https://github.com/vuejs/language-tools/discussions/5456

> create a client handler for vue language server, when receiving the request, forward it to typescript language server and finally send the payload from typescript language server back to vue language server

**Reply by ParsaJR** [reply]
@pbgc Thank you so much. I just didn't noticed the difference between these two versions, in relation with the Eglot capabilities...

**Reply by ParsaJR** [reply]
Hey. Do you encountered any problems regarding to the indentation of vue files? I have this issue for a long time is sfc templates.

Is this because web-mode has limited support for vue files? This must be related to web-mode because the `electric-indent` functioning in emacs is based on the major mode which is web-mode

**Comment by yisraeldov** [comment]
Reviving this thread, What versions are working now with eglot on emacs 30?

**Reply by dabeibao** [reply]
Seems the solutions don't work with volar 3.1.5.
Works again after downgrading to 2.2.2

**Reply by pbgc** [reply]
Yes none of the solutions work with > 3 and will never work (unless eglot changes) because it requires to use a separate lsp for typescript and eglot does not support several at once.

**Reply by mdallastella** [reply]
> Seems the solutions don't work with volar 3.1.5. Works again after downgrading to 2.2.2

It also works with version 2.2.8.

**Reply by dabeibao** [reply]
Searched the web and found this:
https://www.reddit.com/r/emacs/comments/1mj8fjc/eglot_is_good/

which links to the discusstion:
https://github.com/joaotavora/eglot/discussions/1429

Does eglot have any plans to support multiple servers?

**Reply by trev-dev** [reply]
The general sentiment from that thread is that Joao doesn't want to write a multiplexer for Eglot as it's out of scope in their mind...and that's fine. We can somehow try to contribute an implementation or use LSPx or lsp-mode. Lsp-mode currently works. I'm running 3 servers per vue buffer. It _does not work well_.

**Reply by yisraeldov** [reply]
What about the other LSP modes ?

https://github.com/zbelial/lspce

?

( @trev-dev , HI BTW )

**Reply by garyo** [reply]
lspce docs: 
> lspce does not support running multiple servers in a single buffer

**Reply by pbgc** [reply]
I dived a little more on the problem and it's not enough that eglot support multiple servers!

https://github.com/vuejs/language-tools/discussions/5456

The client needs to implement additional steps: "create a client handler for vue language server, when receiving the request, forward it to typescript language server and finally send the payload from typescript language server back to vue language server"

**Reply by trev-dev** [reply]
Nice find @pbgc. The simplicity of a "standard implementation for all languages" is completely lost on VueJS. It's disappointing to learn that other web toolchains have the same problem: breaking that standard. Given how quickly the one working implementation I could find in `lsp-mode` breaks, I'm tempted to just run ESLint and call it a day.
